controllers/helloworld_controller.go

```diff
@@ -0,0 +1,221 @@
+type HelloworldReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cache.my.domain,resources=helloworlds,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cache.my.domain,resources=helloworlds/status,verbs=get;update;patch;create
+//+kubebuilder:rbac:groups=cache.my.domain,resources=helloworlds/finalizers,verbs=update
+
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
+
+func (r *HelloworldReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	Logging(req, "Starting reconcile")
+
+	helloworld := &v1alpha1.Helloworld{}
+
+	if err := r.Get(ctx, req.NamespacedName, helloworld); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		log.Log.Error(err, "unable to fetch Helloworld")
+		return ctrl.Result{}, err
+	}
+
+	stopReconcile, result, err := r.validateOrCreateSvcFor(helloworld, ctx, req)
+
+	if stopReconcile {
+		return result, err
+	}
+
+	stopReconcile, result, err = r.validateOrCreatePodsFor(helloworld, ctx, req)
+
+	if stopReconcile {
+		return result, err
+	}
+
+	Logging(req, "Finished succesfully")
+
+	return ctrl.Result{Requeue: false}, nil
+}
+
+func (r *HelloworldReconciler) validateOrCreateSvcFor(helloworld *v1alpha1.Helloworld, ctx context.Context, req ctrl.Request) (bool, ctrl.Result, error) {
+	svc := &v1.Service{}
+	svcNamespacedName := types.NamespacedName{Namespace: helloworld.Namespace, Name: helloworld.Name + "-service"}
+	stopReconcile := true
+
+	if err := r.Get(ctx, svcNamespacedName, svc); err != nil {
+		if errors.IsNotFound(err) {
+			Logging(req, "Creating a service")
+			r.Create(ctx, r.defineSvcFor(helloworld))
+			return stopReconcile, ctrl.Result{Requeue: true}, nil
+		}
+		log.Log.Error(err, "unable to fetch "+helloworld.Name+"'s service")
+		return stopReconcile, ctrl.Result{}, err
+	}
+
+	Logging(req, "Service exist - "+svc.Name)
+	return !stopReconcile, ctrl.Result{}, nil
+}
+
+func (r *HelloworldReconciler) defineSvcFor(helloworld *v1alpha1.Helloworld) *v1.Service {
+	lbls := labelsForApp(helloworld.Name)
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      helloworld.Name + "-service",
+			Namespace: helloworld.Namespace,
+			Labels:    lbls,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Protocol:   v1.ProtocolTCP,
+					Port:       8080,
+					TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: 8080},
+				}},
+			Selector: lbls,
+			Type:     v1.ServiceTypeClusterIP,
+		},
+	}
+
+	controllerutil.SetControllerReference(helloworld, svc, r.Scheme)
+	return svc
+}
+
+func (r *HelloworldReconciler) validateOrCreatePodsFor(helloworld *v1alpha1.Helloworld, ctx context.Context, req ctrl.Request) (bool, ctrl.Result, error) {
```

Comment:
Functions should preferably be shorter.
Considering extracting status update to a diffrent function (IMO a bit more readable and easier to unit test).
Clean Code is my guide line :)